server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/cs/factory/DeployedCacheStoreFactory.java

```diff
@@ -35,32 +43,42 @@ public StoreConfiguration processConfiguration(StoreConfiguration storeConfigura
 
-         if(deployedCacheStoreMetadata == null) {
-            throw new IllegalStateException("Could not find Deployed Cache metadata for " + deployedConfiguration.getCustomStoreClassName());
-         }
+         try {
+            DeployedCacheStoreMetadata deployedCacheStoreMetadata = getPromise(deployedConfiguration.getCustomStoreClassName()).get(TIMEOUT_SECONDS, TimeUnit.SECONDS);
 
-         StoreConfigurationBuilder replacedStoreBuilder = replacedBuilder.addStore(deployedCacheStoreMetadata.getStoreBuilderClass());
-         replacedStoreBuilder.fetchPersistentState(deployedConfiguration.fetchPersistentState());
-         replacedStoreBuilder.ignoreModifications(deployedConfiguration.ignoreModifications());
-         replacedStoreBuilder.preload(deployedConfiguration.preload());
-         replacedStoreBuilder.purgeOnStartup(deployedConfiguration.purgeOnStartup());
-         replacedStoreBuilder.shared(deployedConfiguration.shared());
-         replacedStoreBuilder.withProperties(deployedConfiguration.properties());
-         StoreConfiguration replacedConfiguration = (StoreConfiguration) replacedStoreBuilder.create();
+            StoreConfigurationBuilder replacedStoreBuilder = replacedBuilder.addStore(deployedCacheStoreMetadata.getStoreBuilderClass());
+            replacedStoreBuilder.fetchPersistentState(deployedConfiguration.fetchPersistentState());
+            replacedStoreBuilder.ignoreModifications(deployedConfiguration.ignoreModifications());
+            replacedStoreBuilder.preload(deployedConfiguration.preload());
+            replacedStoreBuilder.purgeOnStartup(deployedConfiguration.purgeOnStartup());
+            replacedStoreBuilder.shared(deployedConfiguration.shared());
+            replacedStoreBuilder.withProperties(deployedConfiguration.properties());
+            StoreConfiguration replacedConfiguration = (StoreConfiguration) replacedStoreBuilder.create();
 
-         return replacedConfiguration;
+            return replacedConfiguration;
+         } catch (Exception e) {
+            InfinispanLogger.ROOT_LOGGER.error("Could not get Custom Cache Store metadata (" + deployedConfiguration.getCustomStoreClassName() + ") within given time.", e);
+         }
       }
       return null;
    }
 
    public void addInstance(Object instance) {
       DeployedCacheStoreMetadata deployedCacheStoreMetadata = DeployedCacheStoreMetadata.fromDeployedStoreInstance(instance);
-      deployedCacheStores.put(deployedCacheStoreMetadata.getDeployedCacheClassName(), deployedCacheStoreMetadata);
+      getPromise(deployedCacheStoreMetadata.getDeployedCacheClassName()).complete(deployedCacheStoreMetadata);
    }
 
    public void removeInstance(Object instance) {
-      deployedCacheStores.remove(instance.getClass().getName());
+      deployedCacheStores.put(instance.getClass().getName(), new CompletableFuture<>());
```

Comment:
Why are we setting it to new a new empty CompletableFuture?  Is removing not sufficient?